### PR TITLE
[fix] Preserve the real exception (type + stack trace) when a test run aborts in BaseRunTests

### LINKED ISSUE #16161: Repository Quality: Exception Handling and Swallowing Patterns
### 🎯 Repository Quality Improvement Report — Exception Handling and Swallowing Patterns

**Analysis Date**: 2026-06-24
**Focus Area**: Exception Handling and Swallowing Patterns
**Strategy Type**: Custom

### Executive Summary

Analysis of exception handling across the vstest codebase reveals systemic patterns that reduce reliability, cancellation responsiveness, and debuggability. Of 338 `try` blocks in `src/`, 199 (59%) use `catch (Exception)` — the broadest possible catch. Only 2 of these use a specific `catch (OperationCanceledException)` guard, and only 7 use a `when` filter clause. This means cancellation requests triggered by the user can silently be absorbed as generic errors in over 100 code paths.

One confirmed bug was found: `BaseRunTests.cs:221` wraps the caught exception incorrectly, stripping both the original exception type and its full stack trace from the error chain. The codebase has inconsistent OCE handling — `TestRunAttachmentsProcessingManager` sets a good precedent with explicit `catch (OperationCanceledException)` handling, but peer classes like `DiscoveryManager` and `ExecutionManager` lack equivalent protection.

The findings below are prioritized by impact and include exact file locations.

<details>
<summary><b>Full Analysis Report</b></summary>

### Current State Assessment

**Metrics Collected:**

| Metric | Value | Status |
|--------|-------|--------|
| Total `try` blocks in `src/` | 338 | — |
| `catch (Exception)` blocks | 199 (59 % of try blocks) | ⚠️ |
| `catch (OperationCanceledException)` blocks | 2 | ❌ |
| Exception filter (`when`) usages | 7 | ⚠️ |
| `catch (Exception)` without rethrow or OCE guard | 113 | ❌ |
| Try blocks that can swallow `OperationCanceledException` | 101 | ❌ |
| Exception wrapping bugs (`new Exception(msg, ex.InnerException)`) | 1 | ❌ |

#### Strengths
- `TestRunAttachmentsProcessingManager` correctly uses a separate `catch (OperationCanceledException)` before the generic `catch (Exception)`, then checks `operationCanceled.CancellationToken == cancellationToken`.
- `ResultsDirectoryArgumentProcessor.cs` and `LengthPrefixCommunicationChannel.cs` make good use of `when` filters to narrow exception scope.
- `ProxyDiscoveryManager.HandleException` checks `CancellationTokenSource.IsCancellationRequested` to suppress noisy logs on cancellation paths.

#### Areas for Improvement

- **[CRITICAL]** `BaseRunTests.cs:221` — `new Exception(ex.Message, ex.InnerException)` discards the original exception type AND top stack frame. The inner exception should be `ex` (not `ex.InnerException`).
- **[HIGH]** `ParallelOperationManager.DoManagerAction` (line 355) — comment says "can safely ignore" for cancellation scenarios, but the broad `catch (Exception)` swallows ALL exceptions including `OperationCanceledException` from non-cancellation contexts.
- **[HIGH]** `InProcessProxyDiscoveryManager` and `InProcessProxyExecutionManager` — in-process paths (no IPC boundary) have a single `catch (Exception)` covering all failures; OCE from cancelled `CancellationToken` is absorbed.
- **[MEDIUM]** `DataCollectorAttachmentProcessorAppDomain` pipe reader inner loop — catches all exceptions silently, masking pipe errors and preventing clean shutdown on cancellation.
- **[MEDIUM]** 101 `try/catch(Exception)` blocks across the crossplatengine, utilities, and console layers that can silently absorb `OperationCanceledException`.

</details>

---

#### Task 1: Fix exception wrapping bug in `BaseRunTests.RunTests`

**Priority**: High
**Estimated Effort**: Small

`BaseRunTests.cs:221` stores a new `Exception` using `ex.InnerException` as the inner, discarding the original exception's type and stack trace:

```csharp
// Current (BUGGY) — loses exception type and stack trace of ex
exception = new Exception(ex.Message, ex.InnerException);

// Fix — preserve ex as the inner exception
exception = new Exception(ex.Message, ex);
```

This exception is later passed to `RaiseTestRunComplete`, so callers see a generic `Exception` with a broken inner-exception chain instead of the actual failure details (e.g., `SocketException`, `TimeoutException`).

**File**: `src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs`, line 221

---

#### Task 2: Guard `ParallelOperationManager.DoManagerAction` against non-cancellation exceptions

**Priority**: High
**Estimated Effort**: Small

`ParallelOperationManager.DoManagerAction` wraps every parallel action in a bare `catch (Exception)` with a comment that "we can safely ignore" — but this reasoning only holds for cancellation exceptions. All other exceptions (e.g. host-process crash signals, serialization errors) are swallowed here. Use a `when` filter to limit the scope:

```csharp
catch (Exception ex) when (ex is not OperationCanceledException)
{
    EqtTrace.Warning("ParallelOperationManager.DoManagerAction: ...", ex);
}
```

Or, if cancellation truly needs to be silenced too, add an explicit guard and rethrow OCE:

```csharp
catch (OperationCanceledException) { throw; }
catch (Exception ex)
{
    EqtTrace.Warning("ParallelOperationManager.DoManagerAction: ...", ex);
}
```

**File**: `src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelOperationManager.cs`, line 355

---

#### Task 3: Protect in-process discovery and execution from `OperationCanceledException` swallowing

**Priority**: High
**Estimated Effort**: Medium

`InProcessProxyDiscoveryManager` and `InProcessProxyExecutionManager` run test adapters in-process (no IPC boundary). Their single `catch (Exception)` handlers send a discovery/execution-complete message indicating failure — but when a `CancellationToken` fires, the correct outcome is a clean cancellation signal, not an error message. Add an OCE guard following the pattern established by `TestRunAttachmentsProcessingManager`:

```csharp
catch (OperationCanceledException)
{
    // Clean cancellation — don't report as error
    EqtTrace.Info("InProcessProxyDiscoveryManager: Discovery cancelled.");
    // Send aborted-but-not-errored discovery complete
}
catch (Exception exception)
{
    EqtTrace.Error("...", exception);
    // existing error handling
}
```

**Files**:
- `src/Microsoft.TestPlatform.CrossPlatEngine/Client/InProcessProxyDiscoveryManager.cs`, line 68
- `src/Microsoft.TestPlatform.CrossPlatEngine/Client/InProcessProxyexecutionManager.cs`, line 95

---

#### Task 4: Add `OperationCanceledException` guard to `DataCollectorAttachmentProcessorAppDomain` pipe reader

**Priority**: Medium
**Estimated Effort**: Small

The pipe reader loop in `DataCollectorAttachmentProcessorAppDomain` (lines 135 and 143) catches and logs `Exception` broadly. When the cancellation token fires, the pipe stream will throw `OperationCanceledException` which is logged as an error — confusing operators. Add a simple guard:

```csharp
catch (OperationCanceledException)
{
    // Expected on cancellation — exit loop cleanly
    break;
}
catch (Exception ex)
{
    EqtTrace.Error($"...pipe reading...\n{ex}");
}
```

**File**: `src/Microsoft.TestPlatform.CrossPlatEngine/AttachmentsProcessing/DataCollectorAttachmentProcessorAppDomain.cs`, lines 135, 143

---

#### Task 5: Document and enforce the standard exception-handling pattern for vstest

**Priority**: Medium
**Estimated Effort**: Medium

The codebase lacks a consistent strategy for exception handling in async/cancellable code paths. `TestRunAttachmentsProcessingManager` demonstrates the correct pattern but it is not documented or enforced. To prevent regressions:

1. Add a code-style note in `AGENTS.md` (or an `.editorconfig` / analyzer rule if feasible) describing when to use `catch (OperationCanceledException)` vs. `when` filters vs. bare `catch (Exception)`.
2. Consider adding a Roslyn analyzer rule or code review checklist item: any `catch (Exception)` in a path that accepts a `CancellationToken` should either be preceded by `catch (OperationCanceledException)` or use a `when (ex is not OperationCanceledException)` filter.

**Files**: `AGENTS.md`, potentially a custom Roslyn analyzer

---

### 📊 Historical Context

<details>
<summary><b>Previous Focus Areas</b></summary>

| Date | Focus Area | Type |
|------|------------|------|
| 2026-06-22 | TODO/FIXME Technical Debt Triage | Custom |
| 2026-06-23 | Async/Await and Cancellation Robustness | Custom |
| 2026-06-24 | Exception Handling and Swallowing Patterns | Custom |

</details>

---

#### Immediate Actions (This Week)
1. **Fix `BaseRunTests.cs:221` exception wrapping bug** — Priority: High (single-line change, high correctness impact)
2. **Guard `ParallelOperationManager.DoManagerAction`** — Priority: High (two-line change, improves parallel test cancellation)

#### Short-term Actions (This Month)
1. **Add OCE guards to in-process proxy managers** — Priority: High (improves cancellation UX for in-process adapters)
2. **Fix pipe reader OCE handling in `DataCollectorAttachmentProcessorAppDomain`** — Priority: Medium
3. **Document standard exception-handling pattern** — Priority: Medium (prevents future regressions)

*Next analysis: 2026-06-25 — Focus area selected based on diversity algorithm*

> Generated by [Repository Quality Improver](https://github.com/microsoft/vstest/actions/runs/28076343526) · 266.4 AIC · ⌖ 26 AIC · ⊞ 32.5K · [◷](https://github.com/search?q=repo%3Amicrosoft%2Fvstest+is%3Aissue+%22gh-aw-workflow-call-id%3A+microsoft%2Fvstest%2Frepository-quality-improver%22&type=issues)
> - [x] expires  on Jun 26, 2026, 5:10 AM UTC

### PR DESCRIPTION
**Pull request by nohwnd**
When a test run aborts, `BaseRunTests.RunTests` caught the exception and rewrapped it as `new Exception(ex.Message, ex.InnerException)`. That threw away the original exception's **type** and **stack trace** — callers (translation layer, VS Test Explorer) saw a bare `Exception` with a broken (often `null`) inner chain.

The fix preserves the original exception as the inner exception. The one thing we deliberately unwrap is `TargetInvocationException` — the reflection wrapper you get when a test extension is instantiated via `Activator.CreateInstance` and its constructor throws. That wrapper is just noise, so we surface its inner (real) exception. Every other exception is kept as-is. This mirrors the unwrap logic we already use in `test/Intent/Runner.cs`.

Tests cover both paths: a plain exception is preserved as the inner exception, and a `TargetInvocationException` is unwrapped to its real inner exception.

Fixes #16161.

---

> 🔍 *Triaged by [Issue Repro Triage & Auto-Fix 🔍](https://github.com/microsoft/vstest/actions/runs/28140571892)* — refined after review.

### CODE REVIEW

**Review by nohwnd — COMMENTED** [review]
## Code Review — `BaseRunTests.RunTests` Exception Wrapping

**Dimensions activated:** Error Reporting & Diagnostic Clarity, Parallel Execution & Scheduling Safety, IPC Transport & Protocol Stability (via `ExceptionConverter` / `RemoteException` chain)

---

### Fix correctness ✅

The one-line change is correct:

```csharp
// Before — discards ex from the chain entirely
exception = new Exception(ex.Message, ex.InnerException);

// After — preserves ex as InnerException
exception = new Exception(ex.Message, ex);
```

The old code silently severed `ex` from the exception chain. If `ex` had no `InnerException` (the common case), the resulting `TestRunCompleteEventArgs.Error` carried the message but **no stack trace and no type** from the site that actually failed. The fix restores `ex` as the `InnerException`, making the full type, stack trace, and inner chain available.

### IPC serialization impact ✅

I checked `ExceptionConverter.cs` and `RemoteException`. The converter recursively serializes `InnerException`, so after this fix the wire format becomes:

```json
{
  "ClassName": "System.Exception",
  "Message": "...",
  "InnerException": {
    "ClassName": "System.NotImplementedException",
    "StackTraceString": "<full original stack>",
    "InnerException": null
  }
}
```

Before the fix, `InnerException` was `null` (or the wrong sub-chain). This is a **wire-compatible improvement** — `null` → populated is not a breaking change for any reader.

### Test ✅

`RunTestsShouldPreserveOriginalExceptionAsInnerException` correctly captures the `TestRunCompleteEventArgs` from the handler callback and asserts `AreSame(originalException, receivedCompleteArgs.Error!.InnerException)`. It directly validates the fixed invariant. The use of `AreSame` (reference equality) is correct here — instance identity is the right check.

### Minor observation (non-blocking)

The PR description says the original code discards "The original exception type". After this fix, the **top-level** wrapper is still `System.Exception` (not the original type), so type identity at `Error.GetType()` is still lost. The original type is now accessible one level down via `Error.InnerException.GetType()`. This is technically accurate given the description says types are "available to callers", but readers expecting the top-level type to be preserved may be surprised.

This is a pre-existing design choice in the wrapping pattern and is outside the scope of this fix. The improvement to diagnostics is real and meaningful.

---

**No blocking issues found.** The change is minimal, correct, and well-covered by the new test.

🧠 *Reviewed by [Expert Code Reviewer](https://github.com/microsoft/vstest/actions/runs/28141083952)*

> 🧠 *Reviewed by [Expert Code Reviewer 🧠](https://github.com/microsoft/vstest/actions/runs/28141083952)*

**Comment by nohwnd** [comment]
Thanks for the review! No blocking issues found — the fix, test, and IPC serialization impact all look good.

Waiting for Azure DevOps CI to complete. Will iterate if any failures come up.



> 🔧 *Iterated by [PR Iteration Agent 🔧](https://github.com/microsoft/vstest/actions/runs/28141343063)*

File: src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs [hunk 1]
```diff
@@ -218,7 +218,7 @@ public void RunTests()
             {
                 EqtTrace.Error("BaseRunTests.RunTests: Failed to run the tests. Reason: {0}.", ex);
 
-                exception = new Exception(ex.Message, ex.InnerException);
```
Comment (nohwnd) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs:221]: is this maybe done because the RunTestInternal might predominatly throw method invocation exception? and so we unwrap (too agressively), so it woud be better to apply logic like the one we have in GetReal exception, where we unwrap methodInvocationEx but not other exceptions?

Comment (nohwnd) [reply src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs:221]: Good question — I checked where exceptions actually surface to this catch. The per-executor `InvokeExecutor` failures (the reflective test-method path) are caught inside `RunTestInternalWithExecutors`, so they don't reach here; and in the common path executor construction (`_activeExecutor = executor.Value`) is inside that same inner try too. So a `TargetInvocationException` reaching *this* catch is uncommon — but it can still happen (reflective extension instantiation on the debugger-attach path, or adapter/discovery loading inside `GetExecutorUriExtensionMap`).

So I kept neither the blanket `ex.InnerException` unwrap nor the blanket `ex`. Applied the same logic as `test/Intent/Runner.cs`: unwrap `TargetInvocationException` to its inner (real) exception, keep every other exception as-is so its type and stack trace survive. Added a test for the unwrap path; the original bug (dropping type/stack via `ex.InnerException`) stays fixed.

**Review by nohwnd — COMMENTED** [review]
## Re-review — TargetInvocationException unwrap logic (updated commit)

**Dimensions activated:** Error Reporting & Diagnostic Clarity, Parallel Execution & Scheduling Safety, IPC Transport & Protocol Stability

*This is a follow-up review of the updated commit (`c249f42`). The previous expert review covered the original `ex.InnerException → ex` fix. This review focuses on the newly added `TargetInvocationException` unwrapping.*

---

### Updated fix correctness ✅

The revised logic:

```csharp
Exception realException = ex is TargetInvocationException tie && tie.InnerException is not null
    ? tie.InnerException
    : ex;
exception = new Exception(realException.Message, realException);
```

- **Regular exception path**: `realException = ex` — identical to the previous fix, preserves type and stack trace as `InnerException`.
- **TIE unwrap path**: strips the reflection wrapper and surfaces the actual constructor-level exception. Consistent with the pattern in `test/Intent/Runner.cs:78`.
- **Edge case — TIE with null InnerException**: the `is not null` guard correctly falls back to `ex` (the TIE itself), so `exception = new Exception(tie.Message, tie)`. Nothing is lost.

### Can TargetInvocationException actually reach this catch? ✅

I traced the call paths. `RunTestInternalWithExecutors` has its own per-executor `catch (Exception e)` that swallows exceptions from individual executor runs (lines 518–538). What escapes to the outer catch are pre-loop failures — primarily `GetExecutorUriExtensionMap` and extension manager creation. Both go through `Activator.CreateInstance` / reflection, so `TargetInvocationException` is a realistic exception type here. The unwrap is appropriately scoped.

### Test coverage ✅

Both new tests use reference equality (`Assert.AreSame`) — the right check since instance identity is what matters here. The `TargetInvocationException` test correctly verifies that the TIE's inner exception becomes `Error.InnerException` (i.e., the wrapper is fully dropped). The normal exception test verifies the original instance survives the wrap. Both are correct.

### Prior observation still applies (non-blocking)

As noted in the previous review: `Error.GetType()` at the top level is still `System.Exception`, not the original exception type. The real type is accessible via `Error.InnerException.GetType()`. This is a pre-existing design choice in the wrapping pattern and is out of scope for this fix.

### IPC/wire compatibility ✅

Unchanged from the previous review — `null → populated InnerException` is wire-additive and non-breaking for old readers.

---

**No new blocking issues introduced by the TargetInvocationException unwrapping.** The logic is correct, edge-case-safe, and consistent with the codebase convention.

🧠 *Reviewed by [Expert Code Reviewer 🧠](https://github.com/microsoft/vstest/actions/runs/28226551628)*

> 🧠 *Reviewed by [Expert Code Reviewer 🧠](https://github.com/microsoft/vstest/actions/runs/28226551628)*

**Review by azat-msft — APPROVED** [review]
